[Flake] Test_newReadyRacy flaking

https://github.com/kubernetes/kubernetes/pull/116024 added a test that is flaking at a ~3% failure rate with `stress`:

```
go test -race -c k8s.io/apiserver/pkg/storage/cacher
stress ./cacher.test -test.run Test_newReadyRacy
```

_Originally posted by @liggitt in https://github.com/kubernetes/kubernetes/pull/116024#discussion_r1123420776_
            
/sig api-machinery
/assign @aojea
/cc @wojtek-t
/kind failing-test

not sure if it's related, but TestGetCurrentResourceVersionFromStorage in the same package is also flaking:

https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=k8s.io%2Fapiserver%2Fpkg%2Fstorage%2Fcacher

```
{Failed  === RUN   TestGetCurrentResourceVersionFromStorage
panic: Object 'Kind' is missing in '{"name":"pod-1","namespace":"ns","creationTimestamp":null,"Spec":{"RestartPolicy":"","TerminationGracePeriodSeconds":null,"ActiveDeadlineSeconds":null,"NodeSelector":null,"ServiceAccountName":"","NodeName":"","Hostname":"","Subdomain":"","SchedulerName":""},"Status":{"Phase":"","Conditions":null,"Message":"","Reason":"","HostIP":"","PodIP":"","StartTime":null}}
'

goroutine 29741 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.decodeObj({0x2f257b0?, 0xc000142500}, {0x2f28fb0, 0x3fbed20}, {0xc003718900, 0x16c, 0x180}, 0x0?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:458 +0x2af
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*watchChan).prepareObjs(0xc0002061c0, 0xc0034d4ff0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:426 +0x2aa
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*watchChan).transform(0xc0002061c0, 0xc0034d4ff0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:318 +0x65
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*watchChan).processEvent(0xc0002061c0, 0xc0003281b0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:283 +0x1c5
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*watchChan).run
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:161 +0x236
```

--- a/staging/src/k8s.io/apiserver/pkg/storage/cacher/ready_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/cacher/ready_test.go
@@ -18,6 +18,7 @@ package cacher
 
 import (
 	"context"
+	"sync"
 	"testing"
 	"time"
 )
@@ -81,18 +82,26 @@ func Test_newReadyRacy(t *testing.T) {
 	errCh := make(chan error, concurrency)
 	ready := newReady()
 	ready.set(false)
+
+	wg := sync.WaitGroup{}
+	wg.Add(2 * concurrency)
 	for i := 0; i < concurrency; i++ {
 		go func() {
 			errCh <- ready.wait(context.Background())
 		}()
 		go func() {
+			defer wg.Done()
 			ready.set(false)
 		}()
 		go func() {
+			defer wg.Done()
 			ready.set(true)
 		}()
 	}
+	// Last one has to be set to true.
+	wg.Wait()
 	ready.set(true)
+
 	for i := 0; i < concurrency; i++ {
 		if err := <-errCh; err != nil {
 			t.Errorf("unexpected error %v on channel %d", err, i)

